use_cases/eluc/predictors/neural_network/neural_net_predictor.py

```diff
@@ -0,0 +1,297 @@
+            for X, y in train_iter:
+                X, y = X.to(self.device), y.to(self.device)
+                optimizer.zero_grad()
+                out = self.model(X)
+                loss = loss_fn(out.squeeze(), y.squeeze())
+                if log_path:
+                    writer.add_scalar("loss", loss.item(), step)
+                step += 1
+                loss.backward()
+                optimizer.step()
+            
+            # LR Decay
+            if self.step_lr_params:
+                scheduler.step()
+
+            # Evaluate epoch
+            if X_val is not None and y_val is not None:
+                total = 0
+                self.model.eval()
+                with torch.no_grad():
+                    for X, y in tqdm(val_dl):
+                        X, y = X.to(self.device), y.to(self.device)
+                        out = self.model(X)
+                        loss = loss_fn(out.squeeze(), y.squeeze())
+                        total += loss.item() * y.shape[0]
+                
+                if log_path:
+                    writer.add_scalar("val_loss", total / len(val_ds), step)
+                
+                if total < best_loss:
+                    best_model = copy.deepcopy(self.model.state_dict())
+                    best_loss = total
+                    e = time.time()
+                    result_dict["best_epoch"] = epoch
+                    result_dict["best_loss"] = total / len(val_ds)
+                    result_dict["time"] = e - s
+
+                print(f"epoch {epoch} mae {total / len(val_ds)}")
+        
+        if best_model:
+            self.model.load_state_dict(best_model)
+        else:
+            e = time.time()
+            result_dict["time"] = e - s
+
+        # If we provide a test dataset
+        if X_test is not None and y_test is not None:
+            y_pred = self.predict(X_test)
+            y_true = y_test.values
+            mae = np.mean(np.abs(y_pred - y_true))
+            result_dict["test_loss"] = mae
+
+        return result_dict
+
+
+    def predict(self, X_test: pd.DataFrame) -> pd.DataFrame:
+        """
+        Generates prediction from model for given test data.
+        :param X_test: test data to predict on.
+        :return: DataFrame of predictions properly labeled.
+        """
+        X_test_scaled = self.scaler.transform(X_test[self.features])
+        test_ds = CustomDS(X_test_scaled, np.zeros(len(X_test_scaled)))
+        test_dl = DataLoader(test_ds, self.batch_size, shuffle=False)
+        pred_list = []
+        with torch.no_grad():
+            self.model.eval()
+            for X, _ in test_dl:
+                X = X.to(self.device)
+                pred_list.append(self.model(X))
+
+        if len(pred_list) > 1:
+            y_pred = torch.concatenate(pred_list, dim=0).cpu().numpy()
+        else:
+            y_pred = pred_list[0].cpu().numpy()
+        return pd.DataFrame(y_pred, index=X_test.index, columns=[self.label])
```

Comment:
Note: I also copy the index from X_test. I don't believe we do this in ESP because I was having some issues with it. In the ELUC data case where we have a more complicated index (time, lat, lon) rather than just numerically indexed,  this saves the trouble later of having to manually reattach the indices.